File: solutions/stack/alternatives/071-simplify-path.js
/**
 * 071. Simplify Path
 * Medium
 *
 * This problem demonstrates key concepts in Stack.
 *
 * SOLUTION EXPLANATION:
 *
 * INTUITION:
 * Simplify Unix-style file paths by handling '.', '..', and multiple slashes.
 * A stack is perfect for this: push valid directories, pop when encountering '..'.
 *
 * APPROACH:
 * 1. **Split path**: Divide path by '/' to get components
 * 2. **Use stack**: Process each component:
 *    - Skip empty strings and '.'
 *    - Pop from stack for '..' (if stack not empty)
 *    - Push valid directory names
 * 3. **Build result**: Join stack with '/' and prefix with '/'
 *
 * WHY THIS WORKS:
 * - Stack naturally handles directory traversal (.. goes up one level = pop)
 * - Splitting by '/' handles multiple consecutive slashes automatically
 * - Empty components from split are filtered out
 *
 * TIME COMPLEXITY: O(n) - process each character once
 * SPACE COMPLEXITY: O(n) - stack stores valid path components
 *
 * EXAMPLE WALKTHROUGH:
 * ```
 * Input: "/a/./b/../../c/"
 * Step 1: Split → ["", "a", ".", "b", "..", "..", "c", ""]
 * Step 2: Process → stack: [] → ["a"] → ["a"] → ["a","b"] → ["a"] → [] → ["c"]
 * Output: "/c"
 * ```
 *
 * EDGE CASES:
 * - Root path: "/" returns "/"
 * - Path with only dots: "/../" returns "/"
 * - Multiple slashes: "///" returns "/"
 */

/**
 * Main solution for Problem 071: Simplify Path
 *
 * @param {string} path - Unix-style file path
 * @return {string} - Simplified canonical path
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */
function solve(path) {
    const stack = [];
    const components = path.split('/');

    for (const component of components) {
        // Skip empty strings and current directory
        if (component === '' || component === '.') {
            continue;
        }

        // Go up one directory (if possible)
        if (component === '..') {
            if (stack.length > 0) {
                stack.pop();
            }
        } else {
            // Valid directory name
            stack.push(component);
        }
    }

    // Build canonical path
    return '/' + stack.join('/');
}

/**
 * Test cases for Problem 071: Simplify Path
 */
function testSolution() {
    console.log('Testing 071. Simplify Path');

    // Test case 1: Basic with dots and double dots
    const result1 = solve("/a/./b/../../c/");
    const expected1 = "/c";
    console.assert(result1 === expected1, `Test 1 failed: expected ${expected1}, got ${result1}`);

    // Test case 2: Root path
    const result2 = solve("/../");
    const expected2 = "/";
    console.assert(result2 === expected2, `Test 2 failed: expected ${expected2}, got ${result2}`);

    // Test case 3: Multiple slashes
    const result3 = solve("/home//foo/");
    const expected3 = "/home/foo";
    console.assert(result3 === expected3, `Test 3 failed: expected ${expected3}, got ${result3}`);

    // Test case 4: Complex path
    const result4 = solve("/a/../../b/../c//.//");
    const expected4 = "/c";
    console.assert(result4 === expected4, `Test 4 failed: expected ${expected4}, got ${result4}`);

    console.log('All test cases passed for 071. Simplify Path!');
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
    console.log('\n=== Problem 071. Simplify Path ===');
    console.log('Category: Stack');
    console.log('Difficulty: Medium');
    console.log('');

    // Example demonstration would go here
    testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
    demonstrateSolution();
}

// Export for use in other modules
module.exports = {
    solve,
    testSolution,
    demonstrateSolution
};

/**
 * Additional Notes:
 * - This solution focuses on stack concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */

File: docs/solutions/stack/071-simplify-path.py
"""
# Difficulty: Medium

# 071. Simplify Path

Given a problem that demonstrates key concepts in Stack.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>[input description]</dd>
<dt>Output:</dt>
<dd>[output description]</dd>
<dt>Explanation:</dt>
<dd>[explanation]</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>

### INTUITION:
[This problem requires understanding of stack concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply stack methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- The solution leverages stack principles
- Time complexity is optimized for the given constraints
- Space complexity is minimized where possible

### EXAMPLE WALKTHROUGH:
```
Input: [example input]
Step 1: [explain first step]
Step 2: [explain second step]
Output: [expected output]
```

### TIME COMPLEXITY:
O(n)

### SPACE COMPLEXITY:
O(1)

### EDGE CASES:
- Empty input handling
- Single element cases
- Large input considerations

</details>
"""

class Solution:
    def simplifyPath(self, path: str) -> str:
        """
        Simplifies a Unix-style file path to its canonical form.
        
        Args:
            path (str): The file path to simplify
            
        Returns:
            str: The simplified canonical path
            
        Examples:
            >>> solution = Solution()
            >>> solution.simplifyPath("/home//foo/")
            "/home/foo"
            >>> solution.simplifyPath("/a/./b/../../c/")
            "/c"
        """
        # Split the path into components
        components = path.split('/')
        
        # Stack to keep track of valid path components
        stack = []
        
        # Process each component
        for component in components:
            # Skip empty components and current directory (.)
            if not component or component == '.':
                continue
                
            # Handle parent directory (..)
            elif component == '..':
                # If stack is not empty, pop the last directory
                if stack:
                    stack.pop()
            
            # Add valid directory names to stack
            else:
                stack.append(component)
        
        # Construct the canonical path
        # If stack is empty, return root directory
        if not stack:
            return "/"
        
        # Join components with forward slash and add leading slash
        return "/" + "/".join(stack)

def test_solution():
    """
    Test cases for 071. Simplify Path.
    """
    solution = Solution()

    # Test case 1: Basic functionality
    # result = solution.solve([test_input])
    # expected = [expected_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    # Test case 2: Edge case
    # result = solution.solve([edge_case_input])
    # expected = [edge_case_output]
    # assert result == expected, f"Expected {expected}, got {result}"

    print("All test cases passed!")

if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 071. Simplify Path")
